### PR TITLE
[mlir][emitc] Add EmitC lowering for arith.trunci, arith.extsi, arith.extui

### PR DESCRIPTION
**Pull request by cferry-AMD**
These operations can be lowered to EmitC provided the sign-extension and truncation behavior is respected.

Per [C++ Reference](https://en.cppreference.com/w/cpp/language/implicit_conversion#Numeric_conversions): when casting to a narrower integer, truncation is guaranteed if unsigned casts are performed, or C++20 is used regardless of the sign. This implementation sticks to unsigned for trunci, so C++20 is not necessary.

This implementation is a bit more generic than needed by these three operations to accomodate `index_cast` and `index_castui` at a later point (specific `emitc.size_t` and `emitc.ssize_t` types are being discussed).

### CODE REVIEW

**Comment by cferry-AMD** [comment]
Pool of reviewers: @simon-camp @marbre @TinaAMD @mgehre-amd

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 1]
```diff
@@ -112,6 +112,78 @@ class CmpIOpConversion : public OpConversionPattern<arith::CmpIOp> {
...
+  using OpConversionPattern<ArithOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ArithOp op, typename ArithOp::Adaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    Type opReturnType = this->getTypeConverter()->convertType(op.getType());
+    if (!isa_and_nonnull<IntegerType>(opReturnType)) {
+      return rewriter.notifyMatchFailure(op, "expected integer result type");
+    }
```
Comment (TinaAMD) [inline mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:127]:
```suggestion
    if (!isa_and_nonnull<IntegerType>(opReturnType))
      return rewriter.notifyMatchFailure(op, "expected integer result type");
```

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 2]
```diff
@@ -112,6 +112,78 @@ class CmpIOpConversion : public OpConversionPattern<arith::CmpIOp> {
...
+
+    if (adaptor.getOperands().size() != 1) {
+      return rewriter.notifyMatchFailure(
+          op, "CastConversion only supports unary ops");
+    }
```
Comment (TinaAMD) [inline mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:132]:
```suggestion
    if (adaptor.getOperands().size() != 1)
      return rewriter.notifyMatchFailure(
          op, "CastConversion only supports unary ops");
```

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 3]
```diff
@@ -112,6 +112,78 @@ class CmpIOpConversion : public OpConversionPattern<arith::CmpIOp> {
...
+
+    Type operandType = adaptor.getIn().getType();
+    if (!isa_and_nonnull<IntegerType>(operandType)) {
+      return rewriter.notifyMatchFailure(op, "expected integer operand type");
+    }
```
Comment (TinaAMD) [inline mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:137]:
```suggestion
    if (!isa_and_nonnull<IntegerType>(operandType))
      return rewriter.notifyMatchFailure(op, "expected integer operand type");
```

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 4]
```diff
@@ -112,6 +112,78 @@ class CmpIOpConversion : public OpConversionPattern<arith::CmpIOp> {
...
+
+    bool isTruncation = operandType.getIntOrFloatBitWidth() >
+                        opReturnType.getIntOrFloatBitWidth();
+    bool doUnsigned = needsUnsigned || isTruncation;
+
+    Type castType = opReturnType;
+    // For int conversions: if the op is a ui variant and the type wanted as
```
Comment (TinaAMD) [inline mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:144]: Isn't this only supporting int conversions?
```suggestion
    // If the op is a ui variant and the type wanted as
```

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 5]
```diff
@@ -112,6 +112,78 @@ class CmpIOpConversion : public OpConversionPattern<arith::CmpIOp> {
   }
 };
 
+template <typename ArithOp, bool needsUnsigned>
```
Comment (TinaAMD) [inline mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:115]: What does `needsUnsigned` stand for? Does it mean the input should be interpreted as unsigned?

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 6]
```diff
@@ -112,6 +112,78 @@ class CmpIOpConversion : public OpConversionPattern<arith::CmpIOp> {
...
+    bool doUnsigned = needsUnsigned || isTruncation;
+
+    Type castType = opReturnType;
+    // For int conversions: if the op is a ui variant and the type wanted as
+    // return type isn't unsigned, we need to issue an unsigned type to do
+    // the conversion.
+    if (castType.isUnsignedInteger() != doUnsigned) {
+      castType = rewriter.getIntegerType(opReturnType.getIntOrFloatBitWidth(),
+                                         /*isSigned=*/!doUnsigned);
+    }
```
Comment (TinaAMD) [inline mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:150]:
```suggestion
    if (castType.isUnsignedInteger() != doUnsigned)
      castType = rewriter.getIntegerType(opReturnType.getIntOrFloatBitWidth(),
                                         /*isSigned=*/!doUnsigned);
```

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 7]
```diff
@@ -112,6 +112,78 @@ class CmpIOpConversion : public OpConversionPattern<arith::CmpIOp> {
...
+
+    auto result = rewriter.template create<emitc::CastOp>(op.getLoc(), castType,
+                                                          actualOp);
+
+    // Fix the signedness of what this operation returns (for integers,
+    // the arith ops want signless results)
+    if (castType != opReturnType) {
+      result = rewriter.template create<emitc::CastOp>(op.getLoc(),
+                                                       opReturnType, result);
+    }
```
Comment (TinaAMD) [inline mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:170]:
```suggestion
    if (castType != opReturnType)
      result = rewriter.template create<emitc::CastOp>(op.getLoc(),
                                                       opReturnType, result);
```

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 8]
```diff
@@ -112,6 +112,78 @@ class CmpIOpConversion : public OpConversionPattern<arith::CmpIOp> {
...
+    // For int conversions: if the op is a ui variant and the type wanted as
+    // return type isn't unsigned, we need to issue an unsigned type to do
+    // the conversion.
+    if (castType.isUnsignedInteger() != doUnsigned) {
+      castType = rewriter.getIntegerType(opReturnType.getIntOrFloatBitWidth(),
+                                         /*isSigned=*/!doUnsigned);
+    }
+
+    Value actualOp = adaptor.getIn();
+    // Fix the signedness of the operand if necessary
```
Comment (TinaAMD) [inline mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:153]:
```suggestion
    // Adapt the signedness of the operand if necessary
```

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 9]
```diff
@@ -112,6 +112,78 @@ class CmpIOpConversion : public OpConversionPattern<arith::CmpIOp> {
...
+                                  /*isSigned=*/!doUnsigned);
+      actualOp = rewriter.template create<emitc::CastOp>(
+          op.getLoc(), correctSignednessType, actualOp);
+    }
+
+    auto result = rewriter.template create<emitc::CastOp>(op.getLoc(), castType,
+                                                          actualOp);
+
+    // Fix the signedness of what this operation returns (for integers,
+    // the arith ops want signless results)
```
Comment (TinaAMD) [inline mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:166]:
```suggestion
    // Cast to the expected output type
```
Maybe?

File: mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir [hunk 10]
```diff
@@ -177,3 +177,42 @@ func.func @arith_int_to_float_cast_ops(%arg0: i8, %arg1: i64) {
...
+  return %truncd : i8
+}
+
+// -----
+
+func.func @extsi(%arg0: i32) {
+  // CHECK-LABEL: extsi
+  // CHECK-SAME: ([[Arg0:[^ ]*]]: i32)
+  // CHECK: emitc.cast [[Arg0]] : i32 to i64
+
```
Comment (TinaAMD) [inline mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir:200]:
```suggestion
```

File: mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir [hunk 11]
```diff
@@ -177,3 +177,42 @@ func.func @arith_int_to_float_cast_ops(%arg0: i8, %arg1: i64) {
...
+// -----
+
+func.func @extui(%arg0: i32) {
+  // CHECK-LABEL: extui
+  // CHECK-SAME: (%[[Arg0:[^ ]*]]: i32)
+  // CHECK: %[[Conv0:.*]] = emitc.cast %[[Arg0]] : i32 to ui32
+  // CHECK: %[[Conv1:.*]] = emitc.cast %[[Conv0]] : ui32 to ui64
+  // CHECK: emitc.cast %[[Conv1]] : ui64 to i64
+
```
Comment (TinaAMD) [inline mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir:214]:
```suggestion
```

File: mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir [hunk 12]
```diff
@@ -177,3 +177,42 @@ func.func @arith_int_to_float_cast_ops(%arg0: i8, %arg1: i64) {
 
   return
 }
+
```
Comment (TinaAMD) [inline mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir:180]: If seems the other test names in the file start with `aith_`, can you align the new tests to that (also for the unsupported ones)?

**Review by TinaAMD — COMMENTED** [review]
Neat trick with the unsigned interpretation for truncation!

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp (hunk 2 above)
Comment (cferry-AMD) [reply mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:132]: For multi-line statements like this one, I was [asked to put braces](https://github.com/Xilinx/llvm-project/pull/177#discussion_r1593501937). The LLVM [coding standards](https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements) do not give out rules for this specific case, but I'd like to make up my mind -- is there an LLVM officially accepted practice on this?

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp (hunk 4 above)
Comment (cferry-AMD) [reply mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:144]: That's a remnant from a putative index type implementation, that was not subject to this conversion. Good spot!

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp (hunk 5 above)
Comment (cferry-AMD) [reply mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:115]: `needsUnsigned` means that the cast op is to an unsigned type (non-sign-extending, guaranteed truncation prior to C++20). Perhaps `castToUnsigned` would be a better job :)

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp (hunk 9 above)
Comment (cferry-AMD) [reply mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:166]: It's much clearer.

File: mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir (hunk 12 above)
Comment (cferry-AMD) [reply mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir:180]: Done!

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp (hunk 2 above)
Comment (TinaAMD) [reply mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:131]: I don't have strong feelings about that, it seems to be borderline for readability, so I am fine with keeping the braces.

File: mlir/test/Conversion/ArithToEmitC/arith-to-emitc-unsupported.mlir [hunk 13]
```diff
@@ -63,3 +63,22 @@ func.func @arith_cast_fptoui_i1(%arg0: f32) -> i1 {
   return %t: i1
 }
 
+// -----
+
+func.func @arith_index_cast(%arg0: i32) -> i32 {
```
Comment (simon-camp) [inline mlir/test/Conversion/ArithToEmitC/arith-to-emitc-unsupported.mlir:68]: These two tests seem unrelated.

**Review by simon-camp — CHANGES_REQUESTED** [review]
I think truncation to i1 needs to be handled specially, as the arith dialect discards the high bits, but a conversion to bool is similar to `x != 0`. For the same reason signed extension from i1 should be rejected by this pattern. Unsigned extension from i1 works correctly I think.

https://godbolt.org/z/sEcdnz4s4

**Comment by cferry-AMD** [comment]
Yes, the i1 case is special indeed -- thanks for the remark! I added its handling.

File: mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir [hunk 14]
```diff
@@ -177,3 +177,66 @@ func.func @arith_int_to_float_cast_ops(%arg0: i8, %arg1: i64) {
...
+func.func @arith_extui_i1_to_i32(%arg0: i1) {
+  // CHECK-LABEL: arith_extui_i1_to_i32
+  // CHECK-SAME: (%[[Arg0:[^ ]*]]: i1)
+  // CHECK: %[[Conv0:.*]] = emitc.cast %[[Arg0]] : i1 to ui1
+  // CHECK: %[[Conv1:.*]] = emitc.cast %[[Conv0]] : ui1 to ui32
+  // CHECK: emitc.cast %[[Conv1]] : ui32 to i32
+  %idx = arith.extui %arg0 : i1 to i32
+  return
+}
```
Comment (simon-camp) [inline mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir:242]: missing newline

**Review by simon-camp — APPROVED** [review]
Other than the formatting and missing newline this looks good to me.

Out of curiosity, what would be a good way of implementing the sign extension for i1? Doing unsigned extension, oring with `0b1111...00...` and casting back to signed?

File: mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir (hunk 14 above)
Comment (cferry-AMD) [reply mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir]: Solved :)

**Comment by cferry-AMD** [comment]
I would do just the unsigned extension: I don't think it makes sense to interpret the only bit of an i1 as a sign bit (then what's its value?)... then the other choice would be to see a 1 as `+1`, which needs no sign extension to be preserved. 

Now assuming we still sign-extend: as an alternative to the `or` you suggest, we could go unsigned to i8, do `x | 0x7E`, and then perform regular sign-extension with a cast. That would let us work e.g. with index types, which bitwidth is unspecified.
